v2/datastream-to-spanner/terraform/samples/mysql-sharded-end-to-end/README.md

```diff
@@ -0,0 +1,689 @@
+> They will have to be added manually. Note that if they are not added, **the
+> migration will fail.**
+> Two service accounts will need to be modified manually -
+>    1. Dataflow service account - The list of roles can be found in the `main.tf`
+        file, in the `live_migration_roles` resource.
+>    2. GCS service account - The list of roles can be found in the `main.tf` file,
+        in the `gcs_publisher_role` resource.
+>
+>
+>2. In order to create private connectivity configuration for Datastream,
+>`networkAdmin` role is required, [as documented here](https://cloud.google.com/datastream/docs/create-a-private-connectivity-configuration#shared-vpc).
+> Private connectivity cannot be created without these permissions. If you don't want to grant these permissions,
+> you can use the [pre-configured connection profiles template](../pre-configured-conn-profiles/README.md). This template
+> assumes you have created connection profiles outside of Terraform.
+
+[This](#adding-access-to-terraform-service-account) section in the FAQ
+provides instructions to add these roles to an existing service account.
+
+## Assumptions
+
+It takes the following assumptions -
+
+1. MySQL source is accessible via Datastream either via
+   [IP Allowlisting guide](https://cloud.google.com/datastream/docs/network-connectivity-options#ipallowlists)
+   or [Private connectivity](https://cloud.google.com/datastream/docs/create-a-private-connectivity-configuration).
+2. If using a VPC, VPC has already been configured to work with Datastream.
+3. MySQL source has been configured to be read by Datastream by following
+   [configure your source MySQL guide](https://cloud.google.com/datastream/docs/configure-your-source-mysql-database).
+4. A Spanner instance with database containing the data-migration compatible
+   schema is created.
+
+> **_NOTE:_**
+[SMT](https://googlecloudplatform.github.io/spanner-migration-tool/quickstart.html)
+> can be used for converting a MySQL schema to a Spanner compatible schema.
+
+## Description
+
+This sample contains the following files -
+
+1. `main.tf` - This contains the Terraform resources which will be created.
+2. `outputs.tf` - This declares the outputs that will be output as part of
+   running this terraform example.
+3. `variables.tf` - This declares the input variables that are required to
+   configure the resources.
+4. `terraform.tf` - This contains the required providers and APIs/project
+   configurations for this sample.
+5. `terraform.tfvars` - This contains the dummy inputs that need to be populated
+   to run this example.
+6. `terraform_simple.tfvars` - This contains the minimal list of dummy inputs
+   that need to be populated to run this example.
+
+## Resources Created
+
+Given these assumptions, it uses a supplied source database connection
+configuration and creates the following resources -
+
+1. **Datastream private connection** - If configured, a Datastream private
```

Comment:
Nit: It may help to refer to these steps centrally as they are replicated in most of the guides.

Reply:
I think about that quite a bit actually. Ultimately I wanted these samples to be standalone so that the user does not need to navigate across different docs to learn about a given example. 

I agree that updating will be a bit more cumbersome with my approach. I am hoping some of the automation I have added via Github actions will help with validations. We can continue to improve our validators as well.

Reply:
The validator is good, but we need to the documentation to be in sync also. Offline -  there should be some way to sort of write the docs once have them constructed at build time or so ?

Reply:
Yeah there are templating engines who can help with some of this stuff. Can explore